README.md
# UrBackup Exporter

[![Latest release](https://img.shields.io/github/v/release/ngosang/urbackup-exporter)](https://github.com/ngosang/urbackup-exporter/releases)
[![Docker Pulls](https://img.shields.io/docker/pulls/ngosang/urbackup-exporter)](https://hub.docker.com/r/ngosang/urbackup-exporter/)
[![Donate PayPal](https://img.shields.io/badge/Donate-PayPal-yellow.svg)](https://www.paypal.com/paypalme/diegoheras0xff)
[![Donate Bitcoin](https://img.shields.io/badge/Donate-Bitcoin-f7931a.svg)](https://www.blockchain.com/btc/address/1PmeuXvamFM4hkFAJwNdSBee4pRWQW9i4d)
[![Donate Ethereum](https://img.shields.io/badge/Donate-Ethereum-8c8c8c.svg)](https://www.blockchain.com/eth/address/0x0D1549BbB00926BF3D92c1A8A58695e982f1BE2E)

Prometheus exporter for the [UrBackup](https://www.urbackup.org/) backup system. Tested with UrBackup server 2.4.13 and 2.5.31.

Inspired by [h3po work](https://gist.github.com/h3po/36cab38d2b443c0523c4c9e83203f382).

## Install

### Form source code

Requirements:
 * Python 3
 * [prometheus-client](https://github.com/prometheus/client_python)
 * [urbackup-server-web-api-wrapper](https://github.com/uroni/urbackup-server-python-web-api-wrapper)

```bash
pip install -r /requirements.txt

export URBACKUP_SERVER_URL=http://192.168.1.100:55414/x
python urbackup-exporter.py
```

### Docker

Docker images are available in [GHCR](https://github.com/ngosang/urbackup-exporter/pkgs/container/urbackup-exporter) and [DockerHub](https://hub.docker.com/r/ngosang/urbackup-exporter).

```bash
docker pull ghcr.io/ngosang/urbackup-exporter
or
docker pull ngosang/urbackup-exporter
```

#### Supported Architectures

The architectures supported by this image are:

| Architecture     | Default tag (Debian) | Alpine tag |
| ---------------- | -------------------- | ---------- |
| linux/386        | Yes                  | Yes        |
| linux/amd64      | Yes                  | Yes        |
| linux/arm/v6     | No                   | Yes        |
| linux/arm/v7     | Yes                  | Yes        |
| linux/arm64/v8   | Yes                  | Yes        |
| linux/ppc64le    | Yes                  | Yes        |
| linux/s390x      | Yes                  | Yes        |

#### docker-compose

Compatible with docker-compose v2 schemas:

```yaml
---
version: '2.1'
services:
  urbackup-exporter:
    image: ghcr.io/ngosang/urbackup-exporter
    container_name: urbackup-exporter
    environment:
      - TZ=Europe/Madrid
      - URBACKUP_SERVER_URL=http://<server_host>:55414/x
      - URBACKUP_SERVER_USERNAME=admin
      - URBACKUP_SERVER_PASSWORD=1234
      - EXPORT_CLIENT_BACKUPS=true
    ports:
      - 9554:9554
    restart: unless-stopped
```

#### docker cli

```bash
docker run -d \
  --name=urbackup-exporter \
  -e TZ=Europe/Madrid \
  -e URBACKUP_SERVER_URL=http://<server_host>:55414/x \
  -e URBACKUP_SERVER_USERNAME=admin \
  -e URBACKUP_SERVER_PASSWORD=1234 \
  -e EXPORT_CLIENT_BACKUPS=true \
  -p 9554:9554 \
  --restart unless-stopped \
  ghcr.io/ngosang/urbackup-exporter
```

## Configuration

All configuration is done with environment variables.

- `URBACKUP_SERVER_URL`: UrBackup server URL including host, port and API endpoint. Example: `http://192.168.1.100:55414/x`
- `URBACKUP_SERVER_USERNAME`: (Optional) Username to login in the server. Only required if authorization is enabled. The default is `admin`.
- `URBACKUP_SERVER_PASSWORD`: (Optional) Password to login in the server. Only required if authorization is enabled. The default is `1234`.
- `EXPORT_CLIENT_BACKUPS`: (Optional) Export detailed metrics for each client. This option can generate a lot of metrics if there many configured clients. The default is `true`.
- `LISTEN_PORT`: (Optional) The address the exporter should listen on. The default is `9554`.
- `LISTEN_ADDRESS`: (Optional) The address the exporter should listen on. The default is
   to listen on all addresses.
- `LOG_LEVEL`: (Optional) Log level of the traces. The default is `INFO`.

## Exported metrics

| Name                          | Type    | Description                  |
| ----------------------------- | ------- | ---------------------------- |
| urbackup_client_online        | gauge   | Whether or not the client is answering the server |
| urbackup_client_status        | gauge   | Status number, purpose unknown |
| urbackup_client_lastseen      | gauge   | Timestamp the client was last seen online |
| urbackup_backup_ok            | gauge   | Whether or not the last backup was successful |
| urbackup_backup_issues        | gauge   | Number of issues during the last backup |
| urbackup_backup_lasttime      | gauge   | Timestamp of the last backup |
| urbackup_backup_number_total  | counter | Number of backups |
| urbackup_backup_size_total    | counter | Total size of backups in bytes |

## Prometheus config

Example Prometheus configuration:

```yaml
scrape_configs:
  - job_name: 'urbackup-exporter'
    static_configs:
      - targets: ['192.168.1.100:9554']
```

## Prometheus / Alertmanager rules

Example Prometheus rules for alerting:

```yaml
  - alert: UrBackupOutdated
    # 1209600 = 15 days
    expr: time() - urbackup_backup_lasttime{backup_type="file"} > 1209600
    for: 0m
    labels:
      severity: critical
    annotations:
      summary: UrBackup {{ $labels.client_name }} backup is outdated
      description: UrBackup backup is outdated\n  VALUE = {{ $value }}\n  LABELS = {{ $labels }}

  - alert: UrBackupFailure
    expr: urbackup_backup_ok{backup_type="file"} == 0
    for: 0m
    labels:
      severity: critical
    annotations:
      summary: UrBackup {{ $labels.client_name }} backup failed
      description: UrBackup backup failed\n  VALUE = {{ $value }}\n  LABELS = {{ $labels }}
```

## Grafana dashboard

There is a reference Grafana dashboard in [grafana/grafana_dashboard.json](./grafana/grafana_dashboard.json).

![](./grafana/grafana_dashboard.png)


CHANGELOG.md
# Changelog

## 1.2.0 / 2023/10/20

* Improve UrBackup API error handling
* Update Python dependencies
* Update Docker image with Alpine 3.18 and Python 3.12
* Update Docker image with Debian Bookworm (12) and Python 3.12
* Add docker-compose and documentation to test in local
* Tested with UrBackup 2.5.31

## 1.1.3 / 2022/12/06

* Update dependencies and base Docker image
* Add ppc64le & s390x architectures for Alpine, remove mips64le from Debian
* Update Grafana Dashboard

## 1.1.2 / 2022/01/22

* Avoid re-login in each request

## 1.1.1 / 2022/01/15

* Update dependencies and base Docker image
* Update readme

## 1.1.0 / 2021-10-09

* Improve logger to display datetime and reduce error verbosity. Use `TZ=Europe/Madrid` env var to configure timezone
* Add new env var `LOG_LEVEL=INFO` to set the log level
* Update documentation. Reorder sections and include an example to configure Prometheus.

## 1.0.0 / 2021-07-04

* First release
* Tested with UrBackup server 2.4.13


urbackup-exporter.py
#!/usr/bin/env python3

import logging
import os
import time
import sys

import prometheus_client
import prometheus_client.core
import urbackup_api


class UrBackupCollector(object):
    def __init__(self, server, username, password, export_client_backups):
        self.api = None
        self.server = server
        self.username = username
        self.password = password
        self.export_client_backups = export_client_backups

        # connect or fail
        self.get_status_autologin()

    def collect(self):
        logging.debug("Incoming request")

        common_label_names = [
            "client_name",
            "client_group",
            "client_id",
            "client_version",
            "client_os_version"
        ]

        client_online = prometheus_client.core.GaugeMetricFamily(
            "urbackup_client_online",
            "Whether or not the client is answering the server",
            labels=common_label_names)

        client_status = prometheus_client.core.GaugeMetricFamily(
            "urbackup_client_status",
            "Status number, purpose unknown",
            labels=common_label_names)

        client_lastseen = prometheus_client.core.GaugeMetricFamily(
            "urbackup_client_lastseen",
            "Timestamp the client was last seen online",
            labels=common_label_names)

        backup_ok = prometheus_client.core.GaugeMetricFamily(
            "urbackup_backup_ok",
            "Whether or not the last backup was successful",
            labels=common_label_names + ["backup_type"])

        backup_issues = prometheus_client.core.GaugeMetricFamily(
            "urbackup_backup_issues",
            "Number of issues during the last backup",
            labels=common_label_names + ["backup_type"])

        backup_lasttime = prometheus_client.core.GaugeMetricFamily(
            "urbackup_backup_lasttime",
            "Timestamp of the last backup",
            labels=common_label_names + ["backup_type"])

        backup_number_total = prometheus_client.core.CounterMetricFamily(
            "urbackup_backup_number_total",
            "Number of backups",
            labels=common_label_names + ["backup_type", "archived"])

        backup_size_total = prometheus_client.core.CounterMetricFamily(
            "urbackup_backup_size_total",
            "Total size of backups in bytes",
            labels=common_label_names + ["backup_type", "archived"])

        try:
            api_status = self.get_status_autologin()
        except Exception as e:
            logging.error("Unable to connect to UrBackup Server. Error: %s", str(e))
            return

        for client in api_status:
            common_label_values = [
                client["name"],
                client.get("groupname", "N/A"),
                str(client["id"]),
                client.get("client_version_string", "N/A"),
                client.get("os_version_string", "N/A")
            ]

            client_online.add_metric(common_label_values, float(client["online"]))
            client_status.add_metric(common_label_values, float(client["status"]))
            client_lastseen.add_metric(common_label_values, self.float_or_default(client, "lastseen", 0.0))
            backup_ok.add_metric(common_label_values + ["file"], float(client["file_ok"]))
            backup_ok.add_metric(common_label_values + ["image"], float(client["image_ok"]))
            backup_issues.add_metric(common_label_values + ["file"],
                                     self.float_or_default(client, "last_filebackup_issues", 0.0))
            backup_lasttime.add_metric(common_label_values + ["file"],
                                       self.float_or_default(client, "lastbackup", 0.0))
            backup_lasttime.add_metric(common_label_values + ["image"],
                                       self.float_or_default(client, "lastbackup_image", 0.0))

            if self.export_client_backups:
                count_archived, count_no_archived, size_archived, size_no_archived = self.calc_client_backups(
                    self.api.get_clientbackups(client["id"]))
                backup_number_total.add_metric(common_label_values + ["file", "yes"], count_archived)
                backup_number_total.add_metric(common_label_values + ["file", "no"], count_no_archived)
                backup_size_total.add_metric(common_label_values + ["file", "yes"], size_archived)
                backup_size_total.add_metric(common_label_values + ["file", "no"], size_no_archived)
                count_archived, count_no_archived, size_archived, size_no_archived = self.calc_client_backups(
                    self.api.get_clientimagebackups(client["id"]))
                backup_number_total.add_metric(common_label_values + ["image", "yes"], count_archived)
                backup_number_total.add_metric(common_label_values + ["image", "no"], count_no_archived)
                backup_size_total.add_metric(common_label_values + ["image", "yes"], size_archived)
                backup_size_total.add_metric(common_label_values + ["image", "no"], size_no_archived)

        yield client_online
        yield client_status
        yield client_lastseen
        yield backup_ok
        yield backup_issues
        yield backup_lasttime

        if self.export_client_backups:
            yield backup_number_total
            yield backup_size_total

    def get_status_autologin(self):
        try:
            api_status = self.api.get_status()
            if api_status is None:
                raise Exception("Get Status response is empty, check the log traces.")
        except Exception:
            self.api = urbackup_api.urbackup_server(self.server, self.username, self.password)
            api_status = self.api.get_status()
            if api_status is None:
                raise Exception("Get Status response is empty, check the log traces.")
        return api_status

    @staticmethod
    def calc_client_backups(client_backups):
        count_archived = 0
        size_archived = 0
        size_no_archived = 0
        for client_backup in client_backups:
            if client_backup["archived"] == 1:
                count_archived += 1
                size_archived += client_backup["size_bytes"]
            else:
                size_no_archived += client_backup["size_bytes"]
        count_no_archived = len(client_backups) - count_archived

        return count_archived, count_no_archived, size_archived, size_no_archived

    @staticmethod
    def float_or_default(dict, key, default=0.0):
        try:
            return float(dict.get(key, default))
        except Exception:
            return default


if __name__ == "__main__":
    logging.basicConfig(
        format='%(asctime)s %(levelname)-8s %(message)s',
        level=logging.getLevelName(os.environ.get("LOG_LEVEL", "INFO")),
        datefmt='%Y-%m-%d %H:%M:%S',
        handlers=[
            logging.StreamHandler(sys.stdout)
        ]
    )
    logging.info("Starting UrBackup Prometheus Exporter ...")

    try:
        urbackup_server_url = os.environ["URBACKUP_SERVER_URL"]
    except Exception:
        logging.error("Configuration error. The environment variable URBACKUP_SERVER_URL is mandatory")
        sys.exit(1)

    urbackup_server_username = os.environ.get("URBACKUP_SERVER_USERNAME", "admin")
    urbackup_server_password = os.environ.get("URBACKUP_SERVER_PASSWORD", "1234")
    urbackup_export_client_backups = os.environ.get("EXPORT_CLIENT_BACKUPS", "true").lower() == "true"
    exporter_address = os.environ.get("LISTEN_ADDRESS", "0.0.0.0")
    exporter_port = int(os.environ.get("LISTEN_PORT", 9554))

    prometheus_client.core.REGISTRY.register(UrBackupCollector(
        urbackup_server_url, urbackup_server_username, urbackup_server_password, urbackup_export_client_backups))
    prometheus_client.start_http_server(exporter_port, exporter_address)

    logging.info("Server listening in http://%s:%d/metrics", exporter_address, exporter_port)
    while True:
        time.sleep(60)
